Userland/Libraries/LibSemVer/SemVer.cpp

```diff
@@ -0,0 +1,250 @@
+    if (m_prerelease.bytes().size() > 0) {
+        sb.append('-');
+        sb.append(m_prerelease);
+    }
+    if (m_buildmetadata.bytes().size() > 0) {
+        sb.append('+');
+        sb.append(m_buildmetadata);
+    }
+    return MUST(sb.to_string());
+}
+
+String SemVer::to_string() const
+{
+    return MUST(String::formatted("{}{}{}{}{}{}", m_major, m_mmp_sep, m_minor, m_mmp_sep, m_patch, suffix()));
+}
+
+ReleaseType SemVer::release_type() const
+{
+    auto prerel_identifiers = MUST(MUST(m_prerelease.to_lowercase()).split('.'));
+    auto is_alpha = prerel_identifiers.contains_slow("alpha"_string);
+    auto is_beta = prerel_identifiers.contains_slow("beta"_string);
+
+    if (is_alpha && is_beta) {
+        return ReleaseType::AlphaBeta;
+    }
+
+    if (is_alpha) {
+        return ReleaseType::Alpha;
+    }
+
+    if (is_beta) {
+        return ReleaseType::Beta;
+    }
+
+    if (prerel_identifiers.contains_slow("prerelease"_string)) {
+        return ReleaseType::PreRelease;
+    }
+
+    // strict release candidate check
+    if (prerel_identifiers.contains_slow("rc"_string)) {
+        return ReleaseType::ReleaseCandidate;
+    }
+
+    return ReleaseType::Stable;
+}
+
+ErrorOr<SemVer> SemVer::bump(BumpType type, BumpOpts opt) const
+{
+    String prerelease {};
+    String buildmetadata {};
+
+    switch (opt) {
+    case BumpOpts::KeepSuffix:
+        prerelease = m_prerelease;
+        buildmetadata = m_buildmetadata;
+        break;
+    case BumpOpts::ClearBuildMeta:
+        prerelease = m_prerelease;
+        break;
+    case BumpOpts::ClearPreRelease:
+        buildmetadata = m_buildmetadata;
+        break;
+    default:
+        return Error::from_string_view("Invalid bump option"sv);
+    }
+
+    switch (type) {
+    case BumpType::Major:
+        return SemVer(m_major + 1, m_minor, m_patch, m_mmp_sep, prerelease, buildmetadata);
+    case BumpType::Minor:
+        return SemVer(m_major, m_minor + 1, m_patch, m_mmp_sep, prerelease, buildmetadata);
+    case BumpType::Patch:
+        return SemVer(m_major, m_minor, m_patch + 1, m_mmp_sep, prerelease, buildmetadata);
+    default:
+        return Error::from_string_view("Invalid bump type"sv);
+    }
+}
+
+bool SemVer::is_same(SemVer const& other, CompareType compare_type) const
+{
+    switch (compare_type) {
+    case CompareType::Major:
+        return m_major == other.m_major;
+    case CompareType::Minor:
+        return m_major == other.m_major && m_minor == other.m_minor;
+    case CompareType::Patch:
+        return m_major == other.m_major && m_minor == other.m_minor && m_patch == other.m_patch;
+    default:
+        return to_string() == other.to_string();
+    }
+}
+
+bool SemVer::is_greater_than(SemVer const& other) const { return m_major > other.m_major || m_minor > other.m_minor || m_patch > other.m_patch; }
+
+bool SemVer::is_lesser_than(SemVer const& other) const { return m_major < other.m_major || m_minor < other.m_minor || m_patch < other.m_patch; }
```

Comment:
These should be proper comparison operators. Also, they don't seem to account for the prerelease string and they improperly short-circuit.

Reply:
How would you compare two strings if one is "_greater than_" other in sense comes after?

Reply:
I have revamped the logic of pre-release handling in parsing, comparisons and bumps